tests/schema/tests.py

```diff
@@ -2680,23 +2680,150 @@ def test_unique_together(self):
+            assert_implicit_fk_index(
+                [
+                    (editor.add_index, (Book, negative_index)),
+                    (editor.remove_index, (Book, negative_index)),
+                ],
+                # when DESC ordering is not supported by the DB engine
+                # we create indexes in the ascending order.
+                # So in this case the composite index should generate implicit
+                # index on deletion
+                created=not connection.features.supports_index_column_ordering,
+            )
+
+            lookup_index = Index(F("title__lower"), name=index_name)
+            with register_lookup(CharField, Lower):
+                assert_implicit_fk_index(
+                    [
+                        (editor.add_index, (Book, lookup_index)),
+                        (editor.remove_index, (Book, lookup_index)),
+                    ],
+                    created=False,
+                )
+
+            if connection.features.supports_expression_indexes:
+                expression_index = Index(F("author"), F("title"), name=index_name)
+                assert_implicit_fk_index(
+                    [
+                        (editor.add_index, (Book, expression_index)),
+                        (editor.remove_index, (Book, expression_index)),
+                    ],
+                    created=True,
+                )
+
+                desc_expression_index = Index(
+                    F("author").desc(),
+                    F("title"),
+                    name=index_name,
+                )
+                assert_implicit_fk_index(
+                    [
+                        (editor.add_index, (Book, desc_expression_index)),
+                        (editor.remove_index, (Book, desc_expression_index)),
+                    ],
+                    created=False,
+                )
+
+            constraint_name = "test_constraint"
+            self.assertEqual(Book._meta.constraints, [])
+            unique_constraint = UniqueConstraint(
+                fields=["author", "title"],
+                name=constraint_name,
+            )
+            assert_implicit_fk_index(
+                [
+                    (editor.add_constraint, (Book, unique_constraint)),
+                    (editor.remove_constraint, (Book, unique_constraint)),
+                ],
+                created=True,
+            )
+
+            if connection.features.supports_table_check_constraints:
+                check_constraint = CheckConstraint(
+                    check=Q(title__contains="a"),
+                    name=constraint_name,
+                )
+                assert_implicit_fk_index(
+                    [
+                        (editor.add_constraint, (Book, check_constraint)),
+                        (editor.remove_constraint, (Book, check_constraint)),
+                    ],
+                    created=False,
+                )
```

Comment:
Thanks for moving tests to the `schema`, however I still find them complicated and difficult to understand due to the number of internal hooks, flags, and parameters. They are also fragile to isolation issues as you've put them all in one test method.

This patch only fixes a MySQL-specific caveat so complicated tests are unnecessary. I would leave existing tests untouched and add sth like this:
```python
    def test_meta_unique_constraint_with_fk(self):
        with connection.schema_editor() as editor:
            editor.create_model(Author)
            editor.create_model(Book)
        constraint = UniqueConstraint(
            fields=["author", "title"], name="book_author_title_uniq"
        )
        Book._meta.constraints = [constraint]
        with connection.schema_editor() as editor:
            editor.add_constraint(Book, constraint)
        constraints = self.get_constraints(Book._meta.db_table)
        self.assertIn(constraint.name, constraints)
        Book._meta.constraints = []
        with connection.schema_editor() as editor:
            editor.remove_constraint(Book, constraint)
        constraints = self.get_constraints(Book._meta.db_table)
        self.assertNotIn(constraint.name, constraints)

    def _test_meta_index_with_fk(self, index, removed=True):
        with connection.schema_editor() as editor:
            editor.create_model(Author)
            editor.create_model(Book)
        table = Book._meta.db_table
        Book._meta.indexes = [index]
        with connection.schema_editor() as editor:
            editor.add_index(Book, index)
        self.assertIn(index.name, self.get_constraints(Book._meta.db_table))
        Book._meta.indexes = []
        with connection.schema_editor() as editor:
            editor.remove_index(Book, index)
        indexes = self.get_indexes(Book._meta.db_table)
        self.assertNotIn(index.name, self.get_constraints(Book._meta.db_table))
        
    def test_meta_index_with_fk(self):
        index = Index(fields=["author", "title"], name="book_author_title_idx")
        self._test_meta_index_with_fk(index)
        
    def test_meta_index_with_fk(self):
        index = Index(fields=["-author", "title"], name="book_author_title_idx")
        self._test_meta_index_with_fk(index)
        
    @skipUnlessDBFeature("supports_expression_indexes")
    def test_meta_func_index_with_fk(self):
        index = Index(F("author"), F("title"), name="book_author_title_idx")
        self._test_meta_index_with_fk(index)
        
    @skipUnlessDBFeature("supports_expression_indexes")
    def test_meta_desc_func_index_with_fk(self):
        index = Index(F("author").desc(), F("title"), name="book_author_title_idx")
        self._test_meta_index_with_fk(index)
        
    @skipUnlessDBFeature("supports_expression_indexes")
    def test_meta_func_transform_index_with_fk(self):
        index = Index(F("title__lower"), name="book_title_lower_idx")
        with register_lookup(CharField, Lower):
            self._test_meta_index_with_fk(index, removed=False)
```

Reply:
> This patch only fixes a MySQL-specific caveat so complicated tests are unnecessary.

Do you mean that we shouldn't check the implicit index existence, just check that removing indexes/constraints work?

Reply:
makes sense, I didn't known that it is verified as a bug in MySQL (I thought they consider it as a feature :) )

Reply:
done